### DISCUSSION TITLE
abort operations with AbortController

### DISCUSSION THREAD

**Discussion by jimmywarting** [Automattic/mongoose#11811] [Ideas]
An AbortController can come from multiple places such as a spec compatible whatwg/Request or some server that wraps incoming http server request into a standard whatwg/request with a AbortSignal. they can also be constructed with `new AbortController()` or the new proposed `AbortController.timeout(ms)`

This are useful b/c the can cancel multiple operations simultaneously at once.
calling `abort()` on the controller could close the request connection, stop listening on a other multiple event listener, `addEventListener(name, fn, { signal })` and many more thing at once. this signal could also be forwarded to mongoose operation too, to abort any mongodb operation if say a user decide to close the request/connection 

it would be cool if the `.find()`, and `.cursor()` could take an optional `{ signal }` parameter that abort the hole operation when the signal dispatch an abort event, likewise with `db.connect()` and also aggregation

**Comment by Uzlopak** [comment]
Makes sense. Does mongodb native driver support an AbortController?

**Reply by jimmywarting** [reply]
Hmm, don't know. don't think so...?

**Reply by vkarpov15** [reply]
The MongoDB Node driver doesn't currently support an AbortController. In theory, this could be possible with [`killOp()`](https://www.mongodb.com/docs/manual/reference/method/db.killOp/).

**Comment by accidentalprogrammer** [comment]
Any latest info on this? I would also really like to have this feature.

**Reply by vkarpov15** [reply]
We added support for aborting an `eachAsync()` with AbortController here: https://github.com/Automattic/mongoose/pull/12323 . What operations do you want to be able to cancel @accidentalprogrammer ?

**Reply by accidentalprogrammer** [reply]
@vkarpov15 thanks for that info I'll go through that PR and see if it helps me but just I'll explain my problem just in case I'm looking at it the wrong way.
So we have some long running queries in our system which take a lot of time to return and sometimes user gives a very big time range and moves away from the page even before the result returns but that mongo query keeps executing and hogs the connection pool and sometimes we run out of connections and the the UI is unable to load any data after that until some connection frees up or we restart the mongoose server.
What I was looking into is if I can somehow cancel the mongo query that has been run in the event that the underlying http request is cancelled and that data is no longer needed.

**Reply by vkarpov15** [reply]
There are a few problems with that approach, most notably because there isn't really a good way to tell whether the underlying HTTP request has been cancelled or not. I'd recommend just using maxTimeMS: https://thecodebarbarian.com/slow-trains-in-mongodb-and-nodejs

**Comment by GalacticHypernova** [comment]
@vkarpov15 MongoDB Node.js driver now supports aborts for find/aggregate/countDocuments/findOne https://github.com/mongodb/node-mongodb-native/pull/4364.

This could be a good addition as it would be just passing the cursor to mongodb driver. Would you like me to open a PR for that?

**Reply by vkarpov15** [reply]
@GalacticHypernova sure, feel free to put in a PR.

**Reply by Falsen** [reply]
Any progress on this?

**Reply by GalacticHypernova** [reply]
Hey @Falsen !
Mongoose supports abort controller usage just like mongodb, I also added it to the typing with #15549
